File: articles/maladaptive-personality-and-the-border-between-normal-and-clinical-structure/r/analyze_maladaptive_personality.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_maladaptive_personality_structure.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)
trait_cols <- c("negative_affectivity", "detachment", "antagonism", "disinhibition", "psychoticism", "anankastia")
data$dominant_trait_domain <- trait_cols[max.col(data[, trait_cols], ties.method = "first")]
data$severity_band <- ifelse(data$clinical_severity < 2.5, "lower_severity", ifelse(data$clinical_severity < 4.5, "moderate_severity", "higher_severity"))

context_summary <- aggregate(
  data[, c("functioning_impairment", "maladaptive_trait_burden", "rigidity", "pervasiveness", "clinical_severity", "clinical_liability")],
  by = list(clinical_context = data$clinical_context),
  FUN = mean
)
write.csv(context_summary, file.path(outputs, "r_clinical_context_summary.csv"), row.names = FALSE)

severity_summary <- aggregate(
  data[, c("functioning_impairment", "maladaptive_trait_burden", "rigidity", "pervasiveness", "threshold_zone_indicator")],
  by = list(severity_band = data$severity_band),
  FUN = mean
)
write.csv(severity_summary, file.path(outputs, "r_severity_band_summary.csv"), row.names = FALSE)

corr_cols <- c(trait_cols, "identity_impairment", "self_direction_impairment", "empathy_impairment", "intimacy_impairment", "functioning_impairment", "maladaptive_trait_burden", "clinical_severity", "clinical_liability", "threshold_zone_indicator", "rigidity", "pervasiveness", "contextual_stress", "perceived_support")
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_maladaptive_personality_correlations.csv"))

model_severity <- lm(clinical_severity ~ negative_affectivity + detachment + antagonism + disinhibition + psychoticism + anankastia + functioning_impairment, data = data)
model_liability <- lm(clinical_liability ~ maladaptive_trait_burden + functioning_impairment + rigidity + pervasiveness + contextual_stress + perceived_support, data = data)

capture.output(summary(model_severity), file = file.path(outputs, "r_clinical_severity_model_summary.txt"))
capture.output(summary(model_liability), file = file.path(outputs, "r_clinical_liability_model_summary.txt"))
write.csv(data, file.path(outputs, "r_scored_maladaptive_personality.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")

File: articles/maladaptive-personality-and-the-border-between-normal-and-clinical-structure/python/analyze_maladaptive_personality.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_maladaptive_personality_structure.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

TRAITS = ["negative_affectivity", "detachment", "antagonism", "disinhibition", "psychoticism", "anankastia"]
FUNCTIONING = ["identity_impairment", "self_direction_impairment", "empathy_impairment", "intimacy_impairment"]
BOUNDARY = ["rigidity", "pervasiveness", "contextual_stress", "perceived_support"]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"participant_id", "clinical_context", *TRAITS, *FUNCTIONING, "clinical_severity", "clinical_liability", "threshold_zone_indicator", *BOUNDARY}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

df["dominant_trait_domain"] = df[TRAITS].idxmax(axis=1)
df["severity_band"] = np.select(
    [df["clinical_severity"] < 2.5, df["clinical_severity"].between(2.5, 4.5, inclusive="left"), df["clinical_severity"] >= 4.5],
    ["lower_severity", "moderate_severity", "higher_severity"],
    default="unclassified"
)
df["threshold_status"] = np.where(df["clinical_liability"] >= df["clinical_liability"].quantile(.75), "higher_clinical_liability", "lower_or_moderate_liability")

df.groupby("clinical_context").agg(
    n=("participant_id", "count"),
    functioning_impairment_mean=("functioning_impairment", "mean"),
    trait_burden_mean=("maladaptive_trait_burden", "mean"),
    rigidity_mean=("rigidity", "mean"),
    pervasiveness_mean=("pervasiveness", "mean"),
    clinical_severity_mean=("clinical_severity", "mean"),
    clinical_liability_mean=("clinical_liability", "mean"),
).reset_index().to_csv(OUT / "python_clinical_context_summary.csv", index=False)

df.groupby("severity_band").agg(
    n=("participant_id", "count"),
    trait_burden_mean=("maladaptive_trait_burden", "mean"),
    functioning_impairment_mean=("functioning_impairment", "mean"),
    rigidity_mean=("rigidity", "mean"),
    pervasiveness_mean=("pervasiveness", "mean"),
    threshold_zone_mean=("threshold_zone_indicator", "mean"),
).reset_index().to_csv(OUT / "python_severity_band_summary.csv", index=False)

df.groupby("dominant_trait_domain").agg(
    n=("participant_id", "count"),
    clinical_severity_mean=("clinical_severity", "mean"),
    functioning_impairment_mean=("functioning_impairment", "mean"),
    clinical_liability_mean=("clinical_liability", "mean"),
).reset_index().to_csv(OUT / "python_dominant_trait_domain_summary.csv", index=False)

corr_cols = TRAITS + FUNCTIONING + ["functioning_impairment", "maladaptive_trait_burden", "severity_trait_interaction", "clinical_severity", "clinical_liability", "threshold_zone_indicator"] + BOUNDARY
df[corr_cols].corr().to_csv(OUT / "python_maladaptive_personality_correlations.csv")

models = {
    "clinical_severity": ols(df, "clinical_severity", TRAITS + ["functioning_impairment"]),
    "clinical_liability": ols(df, "clinical_liability", ["maladaptive_trait_burden", "functioning_impairment", "rigidity", "pervasiveness", "contextual_stress", "perceived_support"]),
    "threshold_zone_indicator": ols(df, "threshold_zone_indicator", ["maladaptive_trait_burden", "functioning_impairment", "rigidity", "pervasiveness", "contextual_stress", "perceived_support"]),
    "note": "Synthetic demonstration only; not diagnosis, screening, clinical decision support, or risk scoring."
}
(OUT / "python_maladaptive_personality_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_maladaptive_personality.csv", index=False)
print(f"Wrote outputs to {OUT}")
